torchtune/generation/_generation.py

```diff
@@ -4,28 +4,34 @@
 # This source code is licensed under the BSD-style license found in the
 # LICENSE file in the root directory of this source tree.
 
-from typing import Callable, List, Optional
+from typing import Callable, List, Optional, Tuple
 
 import torch
-from torchtune.modules import TransformerDecoder
+from torchtune.modules.transformer import TransformerDecoder
 
 
-def multinomial_sample_one(probs: torch.Tensor) -> torch.Tensor:
+def multinomial_sample_one(
+    probs: torch.Tensor, q: Optional[torch.Tensor] = None
+) -> torch.Tensor:
     """Samples from a multinomial distribution."""
-    q = torch.empty_like(probs).exponential_(1)
+
+    q = torch.empty_like(probs).exponential_(1) if q is None else q
```

Comment:
Maybe I'm being dumb - why do we need q passed in if we define it here?

Reply:
For use with PPO when we define it outside of `generate_next_token` so we aren't passing in the rng object which breaks compile.

Reply:
this seems generally confusing so maybe it's better to always define it in generate and pass it through to generate_next_token.